File: library/src/main/java/com/example/ruru/library/viewholder/BaseViewHolder.java
package com.example.ruru.library.viewholder;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.zip.Inflater;

/**
 * ViewHolder负责存放每个子View并且使用泛型方法获取View中的数据。
 */
public class BaseViewHolder extends RecyclerView.ViewHolder {

  public View itemView;
  //SparseArray：缓存itemView
  private static SparseArray<View> views;

  public BaseViewHolder(View view) {
    super(view);
    this.itemView = view;
    views = new SparseArray<View>();
  }

  public static BaseViewHolder bindView(ViewGroup viewGroup, int layId) {
    View view = LayoutInflater.from(viewGroup.getContext()).inflate(layId, viewGroup, false);
    return new BaseViewHolder(view);
  }

  public View getView() {
    return itemView;
  }

  public <T extends View> T getView(int viewId) {
    View view = views.get(viewId);
    if (view == null) {
      view = itemView.findViewById(viewId);
      views.put(viewId, view);
    }
    return (T) view;
  }

  public void setText(int viewId, String text) {
    TextView textView = getView(viewId);

    Log.d(getClass().getName(), "setText:textview== " + (textView == null) + " text=" + (text == null));
    //textview为空

    textView.setText(text);
  }


  public void setImageResource(int viewId, int resId) {
    ImageView imageView = getView(viewId);
    imageView.setImageResource(resId);
  }

}

File: library/src/main/java/com/example/ruru/library/viewholder/BaseViewHolder.kt
package com.example.ruru.library.viewholder

import android.icu.lang.UCharacter.GraphemeClusterBreak.T
import android.support.v7.widget.RecyclerView
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

/**
 * ViewHolder负责存放每个子View并且使用泛型方法获取View中的数据。
 * Created by lyr on 2019/12/15
 */
class BaseViewHolder: RecyclerView.ViewHolder {

  var itemView: View? = null

  init {
    this.itemView = view
    views = SparseArray<View>()
  }

  constructor(itemView: View, )


  fun getView(): View {
    return itemView
  }

  fun <T extends View>getView(viewId:Int):T
  {
    View view = views . get (viewId);
    if (view == null) {
      view = itemView.findViewById(viewId);
      views.put(viewId, view);
    }
    return (T) view
  }

  fun setText(viewId: Int, text: String) {
    val textView = getView(viewId)

    textView !!.
      //textview为空

      textView.setText(text)
  }

  fun setImageResource(viewId: Int, resId: Int) {
    var imageView = getView(viewId)
    imageView.setImageResource(resId)
  }

  companion object {
    var views: SparseArray<View>? = null  //SparseArray：缓存itemView

    fun bindView(viewGroup: ViewGroup, layId: Int): BaseViewHolder {
      val view = LayoutInflater.from(viewGroup.context).inflate(layId, viewGroup, false)
      return BaseViewHolder(view)
    }
  }
}